H5I bug fix

jrmainzer opened the pull request:
Repaired synchronization errors in H5I caused by inability to enforce
locking order as H5I was used in H5VL elsewhere in the HDF5 library. Removed code to wrap H5I calls in the global mutex inserted as a temporary fix.

Encountered an odd issue with the use of H5TS_thread_id().  As it appeared to be a wild pointer error somewhere in the API test code, coded around the issue for now by using the Pthreads thread ID call instead.  Retained the H5I_BYPASS_HDF5_TID #define in H5Ipkg.h to turn this work around on and off.  Interestingly, this problem seems to go away with CMake builds.

Also encountered very odd failures in the testmthdf5 test.  Note that this test is only built with CMake.  These issues seem to be related to known issues with static builds and attempted dynamic module loads. In any case, the issue goes away with CMake shared builds.  This issue has to be addressed -- but that will have to be another day.

Matt reported several other bugs in H5I, but I was unable to reproduce once the lock ordering issue was addressed, and shared builds were used.

Tested multi-thread debug and production, thread safe, and single thread with autotools build.  Tested multi-thread debug and production with CMake build.

jrmainzer committed:
Repaired synchroniztion errors in H5I caused by inability to enforce
locking order as H5I was used in H5VL elsewhere in the HDF5 library.
Removed code to wrap H5I calls in the global mutex inserted as a
temporary fix.

Encountered an odd issue with the use of H5TS_thread_id().  As it
appeared to be a wild pointer error somewhere in the API test code,
coded around the issue for now by using the Pthreads thread ID call
instead.  Retained the H5I_BYPASS_HDF5_TID #define in H5Ipkg.h to
turn this work around on and off.  Interestingly, this problem seems
to go away with CMake builds.

Also encountered very odd failures in the testmthdf5 test.  Note that
this test is only built with CMake.  These issues seem to be related
to known issues with static builds and attempted dynamic module loads.
In any case, the issue goes away with CMake shared builds.  This issue
has to be addressed -- but that will have to be another day.

Matt reported several other bugs in H5I, but I was unable to reproduce
once the lock ordering issue was addressed, and shared builds were
used.

Tested multi-thread debug and production, thread safe, and single thread
with autotools build.  Tested multi-thread debug and production with
CMake build.

jrmainzer committed:
Modified H5VL_object_data() to handle NULL pointers in the VOL connector
data structures gracefully.

jrmainzer committed:
Added a free list for instances of H5VL_object_t to the VOL code to address
a seg fault caused by attempts to access an instance of H5VL_object_t after
it had been freed.

For now, entries are held on the free list until shutdown, at which point
they are released to the heap.  Must develop heuristics to allow re-use
and/or discard of entries on the free list during normal operation.

This may be a partial fix as other free lists may be needed for other
structures used by the VOL code.

Reproduction of this seg fault was difficult on my machine.  I believe
I have verified the bug fix, but I may be in error.

Tested mt/debug/shared, and mt/production/shared on Charis using a CMake
build.  Also autotools tests for single thread and MT static.

jrmainzer committed:
Update multithread_test.yml

increased timeout from 1 hour to 2 hours

jrmainzer committed:
Update multithread_test.yml

Increased timeout from 2 hours to 3 hours

jrmainzer committed:
Merge branch '1_14_2_multithread' into 1_14_2_multithread

raylu3 commented:
@jrmainzer 

When I tried to compile this branch on my Mac, I got the following errors:

```
Making all in src
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
  CC       H5VLint.lo
../../hdf5_lifeboat/src/H5VLint.c:2325:14: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
        if ( atomic_fetch_add(&(vol_obj->rc), 1) <= 0 ) {
             ^                ~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:142:43: note: expanded from macro 'atomic_fetch_add'
#define atomic_fetch_add(object, operand) __c11_atomic_fetch_add(object, operand, __ATOMIC_SEQ_CST)
                                          ^                      ~~~~~~
../../hdf5_lifeboat/src/H5VLint.c:2331:13: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
            atomic_fetch_sub(&(vol_obj->rc), 1);
            ^                ~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:145:43: note: expanded from macro 'atomic_fetch_sub'
#define atomic_fetch_sub(object, operand) __c11_atomic_fetch_sub(object, operand, __ATOMIC_SEQ_CST)
                                          ^                      ~~~~~~
../../hdf5_lifeboat/src/H5VLint.c:2370:9: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
        atomic_fetch_sub(&(vol_obj->rc), 1);
        ^                ~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:145:43: note: expanded from macro 'atomic_fetch_sub'
#define atomic_fetch_sub(object, operand) __c11_atomic_fetch_sub(object, operand, __ATOMIC_SEQ_CST)
                                          ^                      ~~~~~~
3 errors generated.
make[2]: *** [H5VLint.lo] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
```

jrmainzer commented:
Hi Ray,

    Interesting.  The code compiles (albeit with warnings) on Linux / gcc.  Lets 
discuss at the technical meeting.

                                                                                Best,

                                                                                John

On Thu, Jun 26, 2025, at 9:08 AM, raylu3 wrote:
> *raylu3* left a comment (LifeboatLLC/hdf5_lifeboat#66) <https://github.com/LifeboatLLC/hdf5_lifeboat/pull/66#issuecomment-3008822121>
> @jrmainzer <https://github.com/jrmainzer>
> 
> When I tried to compile this branch on my Mac, I got the following errors:
> 
> `Making all in src
> /Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
>   CC       H5VLint.lo
> ../../hdf5_lifeboat/src/H5VLint.c:2325:14: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
>         if ( atomic_fetch_add(&(vol_obj->rc), 1) <= 0 ) {
>              ^                ~~~~~~~~~~~~~~
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:142:43: note: expanded from macro 'atomic_fetch_add'
> #define atomic_fetch_add(object, operand) __c11_atomic_fetch_add(object, operand, __ATOMIC_SEQ_CST)
>                                           ^                      ~~~~~~
> ../../hdf5_lifeboat/src/H5VLint.c:2331:13: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
>             atomic_fetch_sub(&(vol_obj->rc), 1);
>             ^                ~~~~~~~~~~~~~~
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:145:43: note: expanded from macro 'atomic_fetch_sub'
> #define atomic_fetch_sub(object, operand) __c11_atomic_fetch_sub(object, operand, __ATOMIC_SEQ_CST)
>                                           ^                      ~~~~~~
> ../../hdf5_lifeboat/src/H5VLint.c:2370:9: error: address argument to atomic operation must be a pointer to non-const _Atomic type ('const _Atomic(size_t) *' invalid)
>         atomic_fetch_sub(&(vol_obj->rc), 1);
>         ^                ~~~~~~~~~~~~~~
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/14.0.3/include/stdatomic.h:145:43: note: expanded from macro 'atomic_fetch_sub'
> #define atomic_fetch_sub(object, operand) __c11_atomic_fetch_sub(object, operand, __ATOMIC_SEQ_CST)
>                                           ^                      ~~~~~~
> 3 errors generated.
> make[2]: *** [H5VLint.lo] Error 1
> make[1]: *** [all] Error 2
> make: *** [all-recursive] Error 1
`
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/LifeboatLLC/hdf5_lifeboat/pull/66#issuecomment-3008822121>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARHCPFALI4RBDSIFMMYLT4D3FQEG7AVCNFSM6AAAAAB6J22YJ2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMBYHAZDEMJSGE>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

jrmainzer committed:
Reworked H5VL_object_data() to address a compilation failure that Ray
ran into on his Mac.  Interestingly, this issue doesn't appear on
my Linux box or the automatic regression tests.

jrmainzer committed:
Pull from personal repository -- probably needed due to direct
edits to multithread_tests.yml
Merge branch '1_14_2_multithread' of ssh://ssh.github.com:443/jrmainzer/hdf5_lifeboat into 1_14_2_multithread
